unixfs/src/walk.rs

```diff
@@ -0,0 +1,1040 @@
+                    Some(Cow::Borrowed(bytes)) if !bytes.is_empty() => bytes,
+                    None | Some(Cow::Borrowed(_)) => &[][..],
+                    _ => unreachable!("never used into_owned"),
+                };
+
+                let (cid, name, depth) = self.next.expect("continued without next");
+                match self.current.as_mut() {
+                    Some(current) => current.as_symlink(cid, &name, depth, metadata),
+                    _ => {
+                        self.current =
+                            Some(InnerEntry::new_root_symlink(cid, metadata, &name, depth));
+                    }
+                }
+
+                let state = if let Some(next) = self.pending.pop() {
+                    State::Unfinished(Self {
+                        current: self.current,
+                        next: Some(next),
+                        pending: self.pending,
+                    })
+                } else {
+                    State::Last(self.current.unwrap())
+                };
+
+                Ok(ContinuedWalk::Symlink(contents, Item::from(state)))
+            }
+        }
+    }
+
+    // TODO: we could easily split a 'static value for a directory or bucket, which would pop all
+    // entries at a single level out to do some parallel walking, though the skipping could already
+    // be used to do that... Maybe we could return the filevisit on Skipped to save user from
+    // re-creating one? How to do the same for directories?
+}
+
+/// Represents what the `Walker` is currently looking at. Converted to `Entry` for public API.
+#[derive(Debug)]
+struct InnerEntry {
+    kind: InnerKind,
+    path: PathBuf,
+    metadata: Metadata,
+    depth: usize,
+}
+
+impl From<InnerEntry> for Metadata {
+    fn from(e: InnerEntry) -> Self {
+        e.metadata
+    }
+}
+
+// FIXME: could simplify roots to optinal cid variants?
+#[derive(Debug)]
+enum InnerKind {
+    /// This is necessarily at the root of the walk
+    RootDirectory,
+    /// This is necessarily at the root of the walk
+    BucketAtRoot,
+    /// This is the metadata containing bucket, for which we have a name
+    RootBucket(Cid),
+    /// This is a sibling to a previous named metadata containing bucket
+    Bucket(Cid),
+    /// Directory on any level except root
+    Directory(Cid),
+    /// File optionally on the root level
+    File(Option<Cid>, Option<FileVisit>, u64),
+    /// Symlink optionally on the root level
+    Symlink(Option<Cid>),
+}
+
+/// Representation of the current item of Walker or the last observed item.
+#[derive(Debug)]
+pub enum Entry<'a> {
+    /// Current item is the root directory (HAMTShard or plain Directory).
+    RootDirectory(&'a Path, &'a Metadata),
+    /// Current item is a continuation of a HAMTShard directory. Only the root HAMTShard will have
+    /// file metadata.
+    Bucket(&'a Cid, &'a Path),
+    /// Current item is a non-root plain directory or a HAMTShard root directory.
```

Comment:
shouldn't this be "or a HAMTShard **non-root** directory."?

Reply:
Good catch!

```suggestion
    /// Current item is a non-root plain Directory or a HAMTShard directory.
```